src/wp-admin/includes/image.php

```diff
@@ -393,79 +439,114 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id ) {
+					unset( $new_sizes[ $size_name ] );
+				}
+			}
+		} else {
+			$image_meta['sizes'] = array();
+		}
 
-		if ( is_wp_error( $rotated ) ) {
-			// TODO: Log errors.
+		if ( empty( $new_sizes ) ) {
+			// Nothing to do...
+			return $image_meta;
 		}
-	}
 
-	if ( method_exists( $editor, 'make_subsize' ) ) {
-		foreach ( $new_sizes as $new_size_name => $new_size_data ) {
-			$new_size_meta = $editor->make_subsize( $new_size_data );
+		/*
+		* Sort the image sub-sizes in order of priority when creating them.
+		* This ensures there is an appropriate sub-size the user can access immediately
+		* even when there was an error and not all sub-sizes were created.
+		*/
+		$priority = array(
+			'medium'       => null,
+			'large'        => null,
+			'thumbnail'    => null,
+			'medium_large' => null,
+		);
+
+		$new_sizes = array_filter( array_merge( $priority, $new_sizes ) );
 
-			if ( is_wp_error( $new_size_meta ) ) {
+		// If stored EXIF data exists, rotate the source image before creating sub-sizes.
+		if ( ! empty( $image_meta['image_meta'] ) ) {
+			$rotated = $editor->maybe_exif_rotate();
+
+			if ( is_wp_error( $rotated ) ) {
 				// TODO: Log errors.
-			} else {
-				// Save the size meta value.
-				$image_meta['sizes'][ $new_size_name ] = $new_size_meta;
-				wp_update_attachment_metadata( $attachment_id, $image_meta );
 			}
 		}
-	} else {
-		// Fall back to `$editor->multi_resize()`.
-		$created_sizes = $editor->multi_resize( $new_sizes );
 
-		if ( ! empty( $created_sizes ) ) {
-			$image_meta['sizes'] = array_merge( $image_meta['sizes'], $created_sizes );
-			wp_update_attachment_metadata( $attachment_id, $image_meta );
+		if ( method_exists( $editor, 'make_subsize' ) ) {
+			foreach ( $new_sizes as $new_size_name => $new_size_data ) {
+				$new_size_meta = $editor->make_subsize( $new_size_data, $output_mime_type );
+
+				if ( is_wp_error( $new_size_meta ) ) {
+					// TODO: Log errors.
+				} else {
+					// Save the default mime size in the 'sizes' meta value.
+					if ( 0 === $mime_index ) {
+						$image_meta['sizes'][ $new_size_name ] = $new_size_meta;
+					}
+
+					// Store the mime type specific sub-size in the 'sources' attribute.
+					$source_meta   = array(
+						'file'     => $new_size_meta['file'],
+						'filesize' => 0,
+					);
+					$file_location = path_join( $dirname, $new_size_meta['file'] );
+					if ( file_exists( $file_location ) ) {
+						$source_meta['filesize'] = wp_filesize( $file_location );
+					}
+					$image_meta['sizes'][ $new_size_name ]['sources'][ $output_mime_type ] = $source_meta;
+					wp_update_attachment_metadata( $attachment_id, $image_meta );
+				}
+			}
+		} else {
+			// Fall back to `$editor->multi_resize()`.
+			$created_sizes = $editor->multi_resize( $new_sizes );
+
+			if ( ! empty( $created_sizes ) ) {
+				$image_meta['sizes'] = array_merge( $image_meta['sizes'], $created_sizes );
+				wp_update_attachment_metadata( $attachment_id, $image_meta );
+			}
 		}
```

Comment:
I think this is not correct, for a few reasons:
* It looks like you didn't update `multi_resize` to support a MIME type. _Maybe_ that's okay given it's a legacy function _but_...
* If we don't update `multi_resize`, we definitely shouldn't call it in the `foreach` loop like it is now. Because it doesn't support custom MIME types, it'll just try creating the original MIME type multiple times.

Either we need to ensure `multi_resize` works for multiple MIME types as well, or we need to move the check around which method to become the outermost condition, and then that legacy method would not support creating additional MIME types.

Reply:
Good point. I don't think we can change `multi_resize` as easily because it is in the base class that plugins may have already extended, potentially leading to conflicts, right? 

I'll take the other approach you suggest, if `make_subsize` is not available, we won't support multiple mime types.

Reply:
resolved in latest version. multiple mime types are only supported if the image engine has `make_subsize`